### DISCUSSION TITLE
Improving Pull Request Merge Velocity

### DISCUSSION THREAD

**Discussion by 0x4007** [ubiquity/.github#88] [General]
A major strategic priority for us is to maximize our velocity for merging in pull requests (with high accuracy, meaning that they achieve the specification without bugs). This is because high performing tech startups ship new features and bug fixes lightning fast ⚡️

This is useful to stay competitive (features) and to retain users (keep them happy with fast bug fixes)

I would like to hear everybody's ideas on how we can improve this because right now it feels that average time is about a couple of weeks per pull request (lots of room for improvement)

Some relevant systems that come to mind:

1. **[Direct incentives for reviewers](https://github.com/orgs/ubiquity/discussions/88#discussioncomment-6786192)** to focus on merging faster with high accuracy (we can explore a lot more direct incentives in this discussion, since this is new territory for anybody)  
2. Automated tests (unit, integration, end to end) via CI
3. Manual QA
4. **[Direct incentives for bounty hunters](https://github.com/orgs/ubiquity/discussions/88#discussioncomment-6786193)** to increase accuracy (we currently have a primitive "debt" system if their issue is reopened, which provides them an incentive to fix it before moving on to the next task)

**Comment by 0x4007** [comment]
# Direct Incentives for Reviewers

**Reply by 0x4007** [reply]
More thought is needed for the following, but I realized that as we grow our review team, we start to bog down the merge speed by finding all sorts of tiny issues. 

I think we should strive to change our culture to turn review tasks into new bounties IF the current version achieves the spec minimally. Code quality is subjective here but it will at least "solve the fix" or "make the new feature" faster by allowing the bounty hunter to focus specifically on what was specified for the associated issue, minimizing scope creep and letting us merge faster. The direct incentive for reviewers to break off tasks into new bounties could be that the comment incentives on the pull request are small, but the issue creator incentive is large. They can effectively convert their comment into a new issue/task, which is what I do on a daily basis; hence why almost all of the issues I create reference a previous comment elsewhere.

Also code cleanup/refactoring/details could be handled by a reviewer to just jump in and clean up before merging. We should figure out a direct incentive for this behavior.

This could lead to apparent speed gains.

**Reply by wannacfuture** [reply]
Hmm, yeah right, its the urgent issue for now.

Tbh, I can't say `more reviewers for each PR` is bad. because it will increase code quality more and also remove more small bugs.

But can direct incentives for pull request reviews will solve this problem?
We know that if we add more comments on the PR we will get more incentives for them. As a result, we will try to find more small bugs to get more incentives, right? lol it will slow down the reviewing time.

What we really need to do is to reduce the duration between PR created time and merged time.
What about incentives for the reviewing time?(also not so sure about this because it can reduce the code quality & accuracy)

My opinion is:
@pavlovcik checks current opened PRs at the end of every day(or week, maybe) then assigns them to appropriate reviewers.
every PR will require only 2 reviewers(code reviewer and QA reviewer or guess one reviewer can handle both of them).

Then, if something goes wrong about the merged PR, the reviewers should work on hotfixes about it.
Or if the merged PR is working well, the incentives for pr reviewers will be great, maybe.

**Reply by 0x4007** [reply]
> But can direct incentives for pull request reviews will solve this problem?
> 
> We know that if we add more comments on the PR we will get more incentives for them. As a result, we will try to find more small bugs to get more incentives, right? lol it will slow down the reviewing time.

Like I alluded to above, if the incentive comes in the form of creating a new issue instead of directly from reviewer comments then only the "good" or approved changes will receive issue creator credit (priority and time are associated with them, thereby funding the tasks)

It will also abstract the new task away from the existing pull request which will help with closing out the existing pull request faster. 

> 
> 
> My opinion is:
> 
> @pavlovcik checks current opened PRs at the end of every day(or week, maybe) then assigns them to appropriate reviewers.
> 

I'm actively looking to remove the dependence on me on GitHub because soon I need to be focusing my attention on Series A fundraising and setting up our fiat banking infrastructure. 

> every PR will require only 2 reviewers(code reviewer and QA reviewer or guess one reviewer can handle both of them).
> 

I've seen some GitHub repository setting/feature which "round robin" requests reviewers to load balance the work which seems interesting to explore. I think this might make more sense when we have a larger team (hopefully soon) because right now I feel that I have a pretty good idea of who is best to review what type of work. 


> 
> Then, if something goes wrong about the merged PR, the reviewers should work on hotfixes about it.
> 

We can make an incentive for the reviewers to bet on the "correct side" this requires more thought but for example, approving pull requests that meet the spec and receiving a delayed payout to verify that the pull request isn't reverted. 

Requesting changes can add a small direct (paid on merge) financial bonus possibly.

**Reply by cuddle-dev** [reply]
`Direct incentives` -- Is it different from the incentive feature we are currently implementing? 

The main concern would be to determine how much we should directly pay to reviewers. we can set it up for a small percentage of the original amount. In this case, we need 2 percent values at least. for both code reviewers and QA. ex; code review: 10%, qa: 5% 

Once we roll up the incentive feature, we can engage the reviewers. but should consider we don't pay much for a lot of conversation history.

**Reply by 0x4007** [reply]
> `Direct incentives` -- Is it different from the incentive feature we are currently implementing? 
> 
> 

Same idea different expression. It's what we are already doing but is not limited to what is implement so far. 


> 
> The main concern would be to determine how much we should directly pay to reviewers. we can set it up for a small percentage of the original amount. In this case, we need 2 percent values at least. for both code reviewers and QA. ex; code review: 10%, qa: 5% 
> 
> 
> 
> Once we roll up the incentive feature, we can engage the reviewers. but should consider we don't pay much for a lot of conversation history.

If incentives are capped to bounty price, we could do comment incentives and then a bigger incentive for approve/request changes (a conclusion to the review) e.g. $10-$25

For a $100 bounty, only a few rounds of reviews per reviewer would be incentivized. At that point the bot can automatically close the pull request with a comment saying that the task was failed because no more reviewer incentives can be provided. However if the reviewers are optimistic on the pull request, they can simply reopen it and complete the review process. 

It can just check every reviewer requested. Reviewers can also technically dismiss/delegate (which makes sense in a team that knows what type of code each other team member specializes in)

Once all reviewers run out for incentives it can then do the automatic close.

**Comment by 0x4007** [comment]
# Direct Incentives for Bounty Hunters

**Comment by sergfeldman** [comment]
@pavlovcik I would suggest automations:

- Analyze which GitHub apps can help us in PR review and setup them for experimentation 

https://github.com/marketplace?type=&verification=&query=Pull+request

- Think about the use of automatic AI code review so that the bounty hunters find as many errors as possible themselves

https://github.com/sturdy-dev/codereview.gpt

https://www.cncf.io/blog/2023/06/06/a-chatgpt-powered-code-reviewer-bot-for-open-source-projects/

**Comment by whilefoo** [comment]
I think we get the initial review quite fast but then it kinda slows down. I think it's also because we have a lot of new contributors that don't know how to do QA, don't know the project and its codebase. 

Because of the async nature and bounty hunters leaving and joining the project, we can't ship new features/bugs as fast as most of the startups that have a strong and stable team.

I think what can really improve our velocity is having really good tests, end-to-end including unit tests covering all the major use cases and as bugs arise they can be added to the tests. This way bounty hunters can be more confident in their work as they can test their code and see if anything breaks meaning they can fix it before we even review it, so it saves time for unnecessary reviews and QA.

After the bounty hunter submits the PR, the tests will run in CI and if they succeed, reviewers can be sure that other features are working and focus on reviewing the code quality and QA can focus on only testing the feature/bug that PR is solving and not other features.

It does however increase the burden on the bounty hunter because they have to additionally write tests, but I think it's still better in the end.

**Comment by rndquu** [comment]
> This is because high performing tech startups ship new features and bug fixes lightning fast

High performing tech startups have dedicated teams while we rely on outside collaborators without deep project understanding.

Overall I agree with whilefoo that "unit testing" culture (when unit testing is a reqiement for any feature) should help because this way there are less chances to break something (for a bounty hunter who sees the project for the 1st time). This "unit testing" culture also eliminates technical debt and in the end saves money for payouts (should we have set unit testing initially we wouldn't have had such [issues](https://github.com/ubiquity/ubiquibot/issues/654)).

**Comment by sergfeldman** [comment]
If unit tests are considered as a suitable solution to increase the PR review velocity, it might me helpful to research automation for unit tests as well.

It would be interesting to know real experience with tools that are positioned as "Automated unit test writing, Auto generated tests".

The examples
- https://www.codium.ai/ (TypeScript is supported, also they have an Open Source PR review tool https://www.codium.ai/pr-agent/)
- https://codepal.ai/unit-tests-writer/typescript 
- https://testent.io/automated-unit-test-writing-for-typescript/

**Comment by cuddle-dev** [comment]
From the conversations, all of us seem to agree that both unit tests and end-to-end tests can speed up the PR review.

**Comment by web4er** [comment]
I believe high-quality automated tests will act as the first and strongest line of defense against bugs. This will improve velocity for sure but it will also help hunters, reviewers, and users sleep in peace at night, again indirectly increasing velocity. 

The nature of ubiquibot is that it is a little extra painful to QA. These are no buttons like a website where you fill username and passed and click login and see if it works or not.

> Code quality is subjective here 

Code quality guidelines can be created by discussion in the team. Placed in readme, these guidelines will save everyone's time. Plus, soon we will hear about not just code quality, but also test quality. 

> Also code cleanup/refactoring/details could be handled by a reviewer to just jump in and clean up before merging.

Most hunters will come back and make the same mistakes again. Free up the team's time to review more. Total amount of work contains more testing than actual coding. Usually, I see testing is already being moved to team. Many hunters don't do QA, some provide absolutely useless QA. And now team also will have to do their clean up. 

First time hunters will have to find hard time in review to deliver good quality in second try. And as they provide good quality in second try, velocity will go up. For this, I have some other ideas but they will probably become obsolete when automated tests kick in. Because there wouldn't be a QA anymore. Hunters will write tests with feature/bug fix. 

- each review that finds a bug, hunter has to submit a fresh and complete QA again. Make sure code was run at least once after making a change.
- Ask hunter to cover maximum use cases. If they miss some we provide them during review and ask them to add to QA.
- QA that hunter provides, can we rename it to tests or manual tests? The problem is, where I am from, developers don't do QA. They do tests. QA and tests need different mindset and many developers who are like me, get confused in start, what is this QA and how to do it.

**Comment by 0x4007** [comment]
> I think we get the initial review quite fast but then it kinda slows down. I think it's also because we have a lot of new contributors that don't know how to do QA, don't know the project and its codebase. 
> 
> 
> 
> Because of the async nature and bounty hunters leaving and joining the project, we can't ship new features/bugs as fast as most of the startups that have a strong and stable team.
> 
> 

In the future we can easily gate off the high priority items based on experience points. This will allow only proven talent to handle the important stuff. 

This will also expose us to a perpetual drip of new talent to potentially recruit. But only on "low priority" tasks.   

> 
> I think what can really improve our velocity is having really good tests, end-to-end including unit tests covering all the major use cases and as bugs arise they can be added to the tests. This way bounty hunters can be more confident in their work as they can test their code and see if anything breaks meaning they can fix it before we even review it, so it saves time for unnecessary reviews and QA.
> 
> 

Cool idea let's implement 

> 
> After the bounty hunter submits the PR, the tests will run in CI and if they succeed, reviewers can be sure that other features are working and focus on reviewing the code quality and QA can focus on only testing the feature/bug that PR is solving and not other features.
> 
> 

Are you sure they run in CI from the pull request? I assumed they needed to be merged in to run or something

> 
> It does however increase the burden on the bounty hunter because they have to additionally write tests, but I think it's still better in the end.

We can simply add to the base rate of the repository to attempt to compensate for this. 

---

I also need to do more research into this idea but in theory I think it's possible to have the following flow for software development: 

1. Conversation/issue specification generation 
2. ChatGPT generates draft unit tests based on the codebase and the issue specification (I've been using [cursor.so](https://cursor.so) which consumes your entire codebase for ChatGPT context)
3. Possibly internal collaborator cleans it up
4. Pull requests need to pass all unit tests 
5. After they pass then manual review is possible

Seems very rigid but leveraging ChatGPT for test writing I think is a missed opportunity currently

**Comment by 0x4007** [comment]
> > This is because high performing tech startups ship new features and bug fixes lightning fast
> 
> 
> 
> High performing tech startups have dedicated teams while we rely on outside collaborators without deep project understanding.
> 
> 

I think we can lean into collaborators who stick around for awhile. We can just maintain the open bounties on low priority items for recruiting purposes.   

> 
> Overall I agree with whilefoo that "unit testing" culture (when unit testing is a reqiement for any feature) should help because this way there are less chances to break something (for a bounty hunter who sees the project for the 1st time). This "unit testing" culture also eliminates technical debt and in the end saves money for payouts (should we have set unit testing initially we wouldn't have had such [issues](https://github.com/ubiquity/ubiquibot/issues/654)).

Can you start implementing this culture during the next couple of weeks? We can have a brief discussion around increasing base rate but my expectation is probably at least a 2x which seems reasonable.

**Reply by 0x4007** [reply]
@rndquu following up on this

**Reply by rndquu** [reply]
I will create a "Technical debt" draft issue (to discuss with the team) for the bot's repo with:
- unit testing
- e2e testing
- code refactoring
- updated QA process in terms that we should stop accept PRs with 0 test coverage

**Reply by 0x4007** [reply]
> updated QA process in terms that we should stop accept PRs with 0 test coverage

Although a bit unreliable, I do like the concept of: in the ubiquity dollar repo it checks if the code coverage has decreased, which implies that the newly pushed code does not have associated unit tests.

**Comment by 0x4007** [comment]
> High performing tech startups have dedicated teams while we rely on outside collaborators without deep project understanding.

My vision is to let the best talent rise to the top (quantitatively) with our leveling system, and churn out the poor contributors. 

For example, certain tasks of higher priority can only be handled by certain levels of contributors. 

With the leveling system, it's easy to implement "dedicated teams" as a continuum, not as a boolean.

**Comment by chris-chris** [comment]
Given the complexity of our current architecture, with many microservices and plugins, implementing unit testing and end-to-end (E2E) testing would have a very positive impact on productivity. I'm also familiar with CI processes and am keen to contribute significantly to this area. :)
